components/director/internal/authenticator/middleware.go

```diff
@@ -0,0 +1,136 @@
+package authenticator
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/kyma-incubator/compass/components/director/internal/tenant"
+	"github.com/kyma-incubator/compass/components/director/pkg/scope"
+
+	"github.com/dgrijalva/jwt-go"
+	"github.com/lestrrat-go/jwx/jwk"
+)
+
+const AuthorizationHeaderKey = "Authorization"
+
+type Authenticator struct {
+	jwksEndpoint        string
+	allowJWTSigningNone bool
+	Jwks                *jwk.Set
+}
+
+func New(jwksEndpoint string, allowJWTSigningNone bool) *Authenticator {
+	return &Authenticator{jwksEndpoint: jwksEndpoint, allowJWTSigningNone: allowJWTSigningNone}
+}
+
+func (a *Authenticator) SynchronizeJWKS() error {
+	jwks, err := FetchJWK(a.jwksEndpoint)
+	if err != nil {
+		return errors.Wrapf(err, "while fetching JWKS from endpoint %s", a.jwksEndpoint)
+	}
+
+	a.Jwks = jwks
+	return nil
+}
+
+// TODO: Remove after https://github.com/kyma-incubator/compass/pull/325
+func (a *Authenticator) TempHandler() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			tenant := r.Header.Get(tenant.TenantHeaderName)
+			if tenant == "" {
+				http.Error(w, "No tenant header", http.StatusBadRequest)
+				return
+			}
+
+			ctx := a.contextWithClaims(r.Context(), Claims{
+				Tenant: tenant,
+			})
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+func (a *Authenticator) Handler() func(next http.Handler) http.Handler {
+	return func(next http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			// TODO: Remove getting tenant from header after implementing https://github.com/kyma-incubator/compass/issues/288
+			if r.Header.Get(tenant.TenantHeaderName) == "" {
+				http.Error(w, "No tenant header", http.StatusBadRequest)
+				return
+			}
+
+			bearerToken, err := a.getBearerToken(r)
+			if err != nil {
+				log.Error(errors.Wrap(err, "while getting token from header"))
+				http.Error(w, err.Error(), http.StatusBadRequest)
+				return
+			}
+
+			claims := Claims{}
+			token, err := jwt.ParseWithClaims(bearerToken, &claims, a.getKeyFunc())
+			if err != nil {
+				wrappedErr := errors.Wrap(err, "while parsing token")
+				log.Error(wrappedErr)
+				http.Error(w, wrappedErr.Error(), http.StatusUnauthorized)
+				return
+			}
+
+			if !token.Valid {
+				http.Error(w, "Invalid token", http.StatusUnauthorized)
+				return
+			}
+
+			ctx := a.contextWithClaims(r.Context(), claims)
+
+			next.ServeHTTP(w, r.WithContext(ctx))
+		})
+	}
+}
+
+func (a *Authenticator) getBearerToken(r *http.Request) (string, error) {
+	reqToken := r.Header.Get(AuthorizationHeaderKey)
+	if reqToken == "" {
+		return "", errors.New("Invalid bearer token")
+	}
+	splitToken := strings.Split(reqToken, "Bearer ")
+	reqToken = splitToken[1]
```

Comment:
can we get here `Index Out of Range` err?

Reply:
Not anymore, I've used `TrimPrefix`